client_plugin/drivers/shared/dockerops/dockerops.go

```diff
@@ -142,13 +175,165 @@ func (d *DockerOps) VolumeRemove(volName string) error {
+			SambaUsername + ";" + SambaUsername,
+		"-u",
+		SambaUsername + ";" + SambaPassword}
+	service.TaskTemplate.ContainerSpec.Args = containerArgs
+
+	// Mount a volume on service containers at mount point "/mount"
+	var mountInfo []swarm.Mount
+	mountInfo = append(mountInfo, swarm.Mount{
+		Type:   swarm.MountType("volume"),
+		Source: internalVolumePrefix + volName,
+		Target: "/mount"})
+	service.TaskTemplate.ContainerSpec.Mounts = mountInfo
+
+	// How many containers of this service should be running at a time?
+	// Service mode can be Replicated or Global
+	var uintContainerNum uint64
+	uintContainerNum = 1
+	numContainers := swarm.ReplicatedService{Replicas: &uintContainerNum}
+	service.Mode = swarm.ServiceMode{Replicated: &numContainers}
+
+	/* Ports that the service wants to expose
+	   * Protocol: Samba operates on TCP
+	   * TargetPort: The port within the container that we wish to expose.
+	                 Port on host VM will get self assigned.
+	*/
+	var exposedPorts []swarm.PortConfig
+	exposedPorts = append(exposedPorts, swarm.PortConfig{
+		Protocol:   swarm.PortConfigProtocolTCP,
+		TargetPort: defaultSambaPort,
+	})
+	// service.EndpointSpec is an input for service create.
+	// It carries the previous data structure as well as Mode.
+	// ResolutionModeVIP indicates that networking uses virtual IPs.
+	service.EndpointSpec = &swarm.EndpointSpec{
+		Mode:  swarm.ResolutionModeVIP,
+		Ports: exposedPorts,
+	}
+
+	//Start the service
+	resp, err := d.Dockerd.ServiceCreate(context.Background(),
+		service, options)
+	if err != nil {
+		log.Warningf("Failed to create file server for volume %s. Reason: %v",
+			volName, err)
+		return 0, "", false
+	}
+
+	// Wait till service container starts
+	ticker := time.NewTicker(checkDuration)
+	defer ticker.Stop()
+	timer := time.NewTimer(sambaRequestTimeout)
+	defer timer.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			log.Infof("Checking status of file server container...")
+			port, isRunning := d.isFileServiceRunning(resp.ID, volName)
+			if isRunning {
+				return int(port), volName + serviceNamePostfix, isRunning
+			}
+		case <-timer.C:
+			log.Warningf("Timeout reached while waiting for file server container for volume %s",
+				volName)
+			return 0, "", false
+		}
+	}
+}
+
+// isFileServiceRunning - Checks if a file service container is running
+// It takes some time from service being up to a container running.
+// This function returns true only if the container is up.
+func (d *DockerOps) isFileServiceRunning(servID string, volName string) (uint32, bool) {
+	var port uint32
+	// Grep the samba service running for this volume using service ID
+	serviceFilters := filters.NewArgs()
+	serviceFilters.Add("id", servID)
+	services, err := d.Dockerd.ServiceList(context.Background(),
+		dockerTypes.ServiceListOptions{Filter: serviceFilters})
+	if err != nil {
+		log.Warningf("Failed to check if file server for volume %s was started. %v", volName, err)
+		return port, false
```

Comment:
Why return "port" during failure, and it seems to me that "port" has not been assigned a value yet.